File: java/src/main/java/problems101_200/RotateArray.java
package problems101_200;

enum RotateArray {
    DEFAULT {
        @Override
        void rotate(int[] nums, int k) {
            if (k == 0 || k % nums.length == 0) {
                return;
            }
            if (k > nums.length) {
                k %= nums.length;
            }
            reverse(nums, 0, nums.length - 1);
            reverse(nums, 0, k - 1);
            reverse(nums, k, nums.length - 1);
        }

        void reverse(int[] nums, int left, int right) {
            while (left < right) {
                int temp = nums[left];
                nums[left] = nums[right];
                nums[right] = temp;
                left++;
                right--;
            }
        }
    };

    abstract void rotate(int[] nums, int k);
}

File: kotlin/src/main/kotlin/problems101_200/RotateArray.kt
package problems101_200

enum class RotateArray {
    DEFAULT {
        override fun rotate(nums: IntArray, k: Int) {
            var k = k

            if (k == 0 || k % nums.size == 0) {
                return
            }
            if (k > nums.size) {
                k %= nums.size
            }
            reverse(nums, 0, nums.size - 1)
            reverse(nums, 0, k - 1)
            reverse(nums, k, nums.size - 1)
        }

        fun reverse(nums: IntArray, left: Int, right: Int) {
            var left = left
            var right = right
            while (left < right) {
                val temp = nums[left]
                nums[left] = nums[right]
                nums[right] = temp
                left++
                right--
            }
        }
    },
    ;

    abstract fun rotate(nums: IntArray, k: Int)
}
